File: ios/Classes/ArgosPlugin.swift
import Flutter
import UIKit

public class ArgosPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private static let methodChannelName = "dev.panoptes.argos/native_capture_method"
    private static let eventChannelName = "dev.panoptes.argos/native_capture"

    // Retain the plugin instance and event channel so they are not deallocated
    // after register(with:) returns. FlutterEventChannel is not retained by the
    // registrar, unlike FlutterMethodChannel via addMethodCallDelegate.
    private static var shared: ArgosPlugin?
    private static var eventChannel: FlutterEventChannel?

    private var registered = false

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = ArgosPlugin()
        shared = instance

        let methodChannel = FlutterMethodChannel(
            name: methodChannelName,
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let channel = FlutterEventChannel(
            name: eventChannelName,
            binaryMessenger: registrar.messenger()
        )
        channel.setStreamHandler(instance)
        eventChannel = channel
    }

    // MARK: - FlutterPlugin

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "enable":
            enableCapture()
            result(nil)
        case "disable":
            disableCapture()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func enableCapture() {
        guard !registered else { return }
        registered = true
        URLProtocol.registerClass(ArgosURLProtocol.self)
    }

    private func disableCapture() {
        guard registered else { return }
        registered = false
        URLProtocol.unregisterClass(ArgosURLProtocol.self)
    }

    // MARK: - FlutterStreamHandler

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        ArgosEventSink.shared.setSink(events)
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        ArgosEventSink.shared.setSink(nil)
        return nil
    }
}

File: android/src/main/kotlin/dev/panoptes/argos/ArgosPlugin.kt
package dev.panoptes.argos

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class ArgosPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {

    companion object {
        private const val METHOD_CHANNEL = "dev.panoptes.argos/native_capture_method"
        private const val EVENT_CHANNEL = "dev.panoptes.argos/native_capture"

        @Volatile
        var eventSink: EventChannel.EventSink? = null
    }

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        eventSink = null
    }

    // MethodCallHandler — Android side auto-registers interceptor at plugin attach;
    // enable/disable from Dart control iOS. Android always active once plugin attached.
    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "enable", "disable" -> result.success(null)
            else -> result.notImplemented()
        }
    }

    // StreamHandler
    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}
